official/recommendation/ncf_main.py

```diff
@@ -0,0 +1,300 @@
+  neg_fname = os.path.join(
+      FLAGS.data_dir, FLAGS.dataset + "-" + constants.TEST_NEG_FILENAME)
+  t1 = time.time()
+  ncf_dataset = dataset.data_preprocessing(
+      train_fname, test_fname, neg_fname, FLAGS.num_neg)
+  tf.logging.info("Data preprocessing: {:.1f} s".format(time.time() - t1))
+
+  # Create NeuMF model and convert it to Estimator
+  tf.logging.info("Creating Estimator from Keras model...")
+  keras_model = neumf_model.NeuMF(
+      ncf_dataset.num_users, ncf_dataset.num_items, FLAGS.num_factors,
+      ast.literal_eval(FLAGS.layers), FLAGS.batch_size, FLAGS.reg_mf)
+  num_gpus = get_num_gpus(FLAGS.num_gpus)
+  est_model = convert_keras_to_estimator(keras_model, num_gpus, FLAGS.model_dir)
+
+  # Training and evaluation cycle
+  def train_input_fn():
+    return dataset.input_fn(
+        True, per_device_batch_size(FLAGS.batch_size, num_gpus),
+        FLAGS.epochs_between_evals, ncf_dataset, FLAGS.num_parallel_calls)
+
+  total_training_cycle = (FLAGS.train_epochs //
+                          FLAGS.epochs_between_evals)
+  for cycle_index in range(total_training_cycle):
+    tf.logging.info("Starting a training cycle: {}/{}".format(
+        cycle_index, total_training_cycle - 1))
+
+    # Train the model
+    train_cycle_begin = time.time()
+    est_model.train(input_fn=train_input_fn,
+                    hooks=[tf.train.ProfilerHook(save_steps=10000)])
+    train_cycle_end = time.time()
+
+    # Evaluate the model
+    eval_cycle_begin = time.time()
+    hr, ndcg = evaluate_model(
+        est_model, FLAGS.batch_size, num_gpus, ncf_dataset.eval_gt_items,
+        ncf_dataset.eval_all_items, FLAGS.num_parallel_calls)
+    eval_cycle_end = time.time()
+
+    # Log the train time, evaluation time, and HR and NDCG results.
+    tf.logging.info(
+        "Iteration {} [{:.1f} s]: HR = {:.4f}, NDCG = {:.4f}, [{:.1f} s]"
+        .format(cycle_index, train_cycle_end - train_cycle_begin, hr, ndcg,
+                eval_cycle_end - eval_cycle_begin))
+
+  # Remove temporary files
+  os.remove(constants.TRAIN_DATA)
+  os.remove(constants.TEST_DATA)
+
+
+if __name__ == "__main__":
+  tf.logging.set_verbosity(tf.logging.INFO)
+  parser = argparse.ArgumentParser()
+
+  parser.add_argument(
+      "--model_dir", nargs="?", default="/tmp/ncf/",
+      help="Model directory.")
+  parser.add_argument(
+      "--data_dir", nargs="?", default="/tmp/movielens-data/",
+      help="Input data directory. Should be the same as downloaded data dir.")
+  parser.add_argument(
+      "--dataset", nargs="?", default="ml-1m", choices=["ml-1m", "ml-20m"],
+      help="Choose a dataset.")
+  parser.add_argument(
+      "--train_epochs", type=int, default=20,
+      help="Number of epochs.")
+  parser.add_argument(
+      "--batch_size", type=int, default=256,
+      help="Batch size.")
+  parser.add_argument(
+      "--num_factors", type=int, default=8,
+      help="Embedding size of MF model.")
```

Comment:
embedding_size maybe? "Factors" is used a lot in MF proper, but our users are more likely deep-learning-first rather than MF-first.